Refactor FIPS & unify CLI actions

adamjanovsky opened the pull request:
## High-level plan

- [x] 1. Unify CLIs into a single CLI (#212)
- [x] 2. Migrate tests to `pytest`, separate download tests from others, refactor test structure (split into cc/fips/others)
    - #76, #79
- [x] 3. Refactor FIPS code
- [x] 4. Misc: update docs, docker, readme, ... 

## Tests
- [x] Add tests that check working pandas serialization
- [x] Allow for default paths of `CPEDataset` and `CVEDataset`
    - ~~Or forbid it and demand explicit paths (that will not get serialized omg). Either way, act consistently.~~
- [x] Get rid of class design in the favour of fiixtures
- [x] Refactor module design
- [x] Test maintenance updates
- [x] Select more appropriate names for the methods
- [x] Migrate to pytest
- [x] Divide logic of tests into download / rest. 
- [x] Allow download test to fail on unreliable resources, ~~test download on resources that we trust (e.g. our server)~~
    - ~~Forbid download on non-download test by auto-use fixture that raises exception on `requests.get` / `requests.post`~~

## FIPS

**Processing steps progress:**

- [x] `get_certs_from_web()`
- [x] `process_auxillary_datasets()`
- [x] `download_all_artifacts()`
- [x] `convert_all_pdfs()`
- [x] `analyze_certificates()`
    - [x] `_compute_references()`
    - [x] `_compute_transitive_vulnerabilities()`

**Random notes to do there:**
- [x] Refactor FIPS to hold auxilary datasets in the designated class
- [x] Implement new folder structure
- [x] Rewrite `redo` -> `fresh`, implement reatempting computations on failed samples
- [x] Introduce InternalState both to `FIPSCertificate` and to `FIPSDataset`

**Done outside of the processing steps:**
- [x] Delete plot graph methods
- ~~[ ] Add examples of plot graph into FIPS reference notebook~~
- [x] **Verify intermediate results on small-scale dataset before proceeding further**
- [x] ~~Verify if FIPS webpages are loading on scroll (historical modules). If yes, figure out how to force full load with BeautifulSoup~~
    - This was not confirmed by manual analysis.
- [x] Implement copying of `FIPSDataset` contents
- [x] Unify CLIs, closes #212 
- [x] `FIPSAlgorithmDataset` should reside in own json in auxillary datasets

> It is unclear what things like clean_cert_ids, or algorithms in the various certicate subobjects (PdfData and WebData) are, are they processed or raw, what sort of cleaning was done and what method needs to be called to do it? Specifically this cleaning that is done on FIPS certs is split between the dataset class and the certificate class in a nasty way, this should be unified and moved, maybe even to a separate class similar to how CC certificate ID canonicalization is separate.

- [x] Address the above

> Incomplete: See parse_html_main, the logic there just is not good.

- [x] Address the above

## Misc
- [x] Refactor CC to hold auxilary datasets in the designated class
- [x] Update docs
- [x] Update Docker image
- [x] Restrict usage of `fresh` on places where it does not makes sense
- [x] Update project readme
- [x] Revisit all TODOs (internal check)
- [x] Unify logging messages.
- [ ] Update code documentation
- [x] PyUpgrade and flake plugin into linting pipeline

## Sanity check

- [x] Perform full FIPS run
- [x] Perform full CC run
- [x] Run all notebooks

## Still part of refactoring, but to-be-addressed by separate PRs

- [x] Move content from examples to notebooks/examples
- [ ]  #287 
- [x] Resolve dependabot issues https://github.com/crocs-muni/sec-certs/security/dependabot
- [x] Across tool, unify `from __future__ import annotations`
- [x] Test GH action yields some warnings about usage of deprecated functions, fix.
- [ ] https://github.com/crocs-muni/sec-certs/issues/231
- [ ] Release

adamjanovsky commented:
>     * Data from the FIPS algorithm dataset is not utilized and mined fully. We can follow the links to the algorithm page and get more data that will help us.

Thanks for the input. This I'll probably leave for future work, this PR is going to be giant anyway. Could you please create a separate issue to address this?

J08nY commented:
> > ```
> > * Data from the FIPS algorithm dataset is not utilized and mined fully. We can follow the links to the algorithm page and get more data that will help us.
> > ```
> 
> Thanks for the input. This I'll probably leave for future work, this PR is going to be giant anyway. Could you please create a separate issue to address this?

Did this in #276.

J08nY commented on sec_certs/dataset/dataset.py:
```diff
@@ -433,3 +434,10 @@ def compute_related_cves(
             f"In total, we identified {n_vulnerabilities} vulnerabilities in {n_vulnerable} vulnerable certificates."
         )
+
+    def get_keywords_df(self) -> pd.DataFrame:
```
I think I would actually keep the original name, because for CC it makes sense to get a similar dataframe for certification report keywords and do analysis on that. I am not sure how to unify this.

adamjanovsky replied:
I attempted to resolve it with https://github.com/crocs-muni/sec-certs/pull/275/commits/1b150aa58b1ff8effb35649201ad590a41d3cc7e

adamjanovsky replied:
I'm afraid that `getattr()` will be our friend in this PR 😆

J08nY replied:
Nice!

adamjanovsky commented:
I'll re-implement the folder structure of FIPS Dataset. Outline:

```
dset
├── FIPS Dataset.json
├── auxillary_datasets
│   └── algorithms
│       └── fips_algorithms.json
├── certs
│   ├── modules
│   └── policies
│       ├── pdf
│       └── txt
└── web
    ├── fips_modules_active.html
    ├── fips_modules_historical.html
    └── fips_modules_revoked.html
```

J08nY commented on sec_certs/sample/fips.py:
```diff
@@ -686,26 +685,27 @@ def extract_algorithm_certificates(current_text):
-                continue
+        pass
+
+    # def _process_to_pop(self, reg_to_match: Pattern, cert: str, to_pop: Set[str]) -> None:
```
Haha I love this. This was actually doing some filtering to remove "cert_id" like matches that are likely not cert_ids but algorithm ids or something else.

J08nY commented on cli.py:
```diff
@@ -161,9 +161,9 @@ def main(
         ProcessingStep(
             "download",
-            "download_all_pdfs",
+            "download_artifacts",
```
This should be `download_all_artifacts` if it is intended to match the method name on the dataset.

adamjanovsky commented on sec_certs/sample/fips.py:
```diff
@@ -244,63 +244,60 @@ class InternalState(ComplexSerializableType):
-            file_status: Optional[bool],
-            txt_state: bool,
+            tables_done: bool = False,
+            file_status: bool = True,  # TODO: Check if this is correct
```
Yea, all this commit is very much WiP, unification will CC status will happen

adamjanovsky commented:
@J08nY looking at the code here and there, I planned the following:

- I decomposed whole processing pipeline into four steps: (i) parse metadata, (ii) download artifacts, (iii) process artifacts, (iv) analyze artifacts
- Each child class would hold a property that would return a list of callable for each step, e.g. `artifact_download_methods`, see https://github.com/crocs-muni/sec-certs/blob/9433658b55c6025567fe4101dae017afc9b4f8a0/sec_certs/dataset/fips.py#L57
- Parent class would just orchestrate these steps, i.e. call all required methods (that must have no arguments which so far holds), see https://github.com/crocs-muni/sec-certs/blob/9433658b55c6025567fe4101dae017afc9b4f8a0/sec_certs/dataset/dataset.py#L234
- I try to introduce the state variables and execute the method on the individual samples in the same fashion as in CC world.

This comes with some advantages:
- Refactoring is very fast, because I just follow the design already implemented in CC
- It should also produce fewer errors, same reasoning as above

Some disadvantages that cross my mind:
- Whatever we're unsatisfied with in CC will probably hit also FIPS 😆 
- There's a lot of spaghetti code that we could possibly get rid of. E.g., see `_download_policies()` and `_download_modules()`: https://github.com/crocs-muni/sec-certs/blob/9433658b55c6025567fe4101dae017afc9b4f8a0/sec_certs/dataset/fips.py#L94, https://github.com/crocs-muni/sec-certs/blob/9433658b55c6025567fe4101dae017afc9b4f8a0/sec_certs/dataset/fips.py#L78. They are very similar, just one variable that changes in multiple places + some logging. The same problem applies to methods executed on the individual samples
- I am not certain whether we want to attempt to solve the problem above. First, I'm not sure how to do that without `getattr("lame string of some variable")`. Second, if we break the pattern, we would need to introduce new design... 

Now's the time to speak up if you have comments/tips/whatever :).

J08nY commented:
Interesting approach but I am worried about a few things:
 - Passing callables around smells. For example "(that must have no arguments which so far holds)" as a limitation sounds pretty bad. Maybe the solution here is to not have the superclass orchestrate but instead have it the other way around? I dunno, but passing callables seems to smell and locks us in a thing that is very constraining.
 - I would maybe have a quick look at design patterns to see if some fits? https://refactoring.guru/design-patterns/catalog

I guess I am really not sure what problem you are trying to solve by passing the callables. I will have to look at the code.

J08nY commented:
Looking at the code a bit. I think now I would just match the FIPS Dataset API to the CC Dataset API manually, even at the cost of some code duplication. I don't feel like it is a good time to rework the handling like this during this FIPS unification.

J08nY commented:
Ah yes, lets do this: https://refactoring.guru/design-patterns/template-method and don't pass callables around. Just call the methods directly, even if there might be some duplication I think it makes sense for future flexibility.

adamjanovsky commented:
I was thinking about the logic of the download testing. At the moment, the situation is as follows:

- Tests that don't test download functionality from unreliable servers (CC, FIPS), but use it (when this is difficult to get rid of) check whether the download succeeded and raise `pytest.xfail` if it did not
- Tests that test download from unreliable servers were decorated with `xfail`. The problem with tested functionality is discovered only on repeated failures.
     - We could possibly re-run these tests, but that would require pytest plugins?
- Tests that download from reliable servers were left as they were.

adamjanovsky commented:
@J08nY Some resolution of the problem that you've pointed to in https://github.com/crocs-muni/sec-certs/pull/275/commits/90521384dc416b8dfb22f12b27e22e2f14ad50d6

J08nY commented on sec_certs/sample/fips.py:
```diff
@@ -459,7 +459,7 @@ def pandas_tuple(self) -> Tuple:
             self.web_data.level,
             self.web_data.embodiment,
-            self.web_data.date_validation[0] if self.web_data.date_validation else np.nan,
+            self.web_data.validation_history[0] if self.web_data.validation_history else np.nan,
```
This should likely have `.date` added after the `[0]` right?

adamjanovsky replied:
Sure, thanks.

adamjanovsky commented:
@J08nY Do you have any idea how the cert_id normalization in FIPS works? I basically finished the refactoring, apart from:
- Cleaning up (I'll do that)
- Cert ID normalization
- Reference graph construction
- Computation of transitive vulnerabilities

IMO, ID of the certificate itself is fully determined by the contents from the module html page. What needs to get cleaned are various cert_ids detected as keywords in various sources: caveat text, pdf of the policy, etc. These point to other certificates, but are noisy and may point to Algorithms, or may just represent abitrary numbers, right?

So the goal, prior to constructing the graph of references, would be to prune the *foreign* cert_ids that are detected in `FIPSCertificate`. Does this make sense? Formally, there's no cert_id normalization or anything like that, you just delete detected foreign cert_ids that you consider weak in a sense. 

Also, I'll probably try to convince @GeorgeFI to take a look at building a graph of references, and also the transitive vulnerabilities. He did this for CC, so it makes sense to let him do this also here. 

What do you think?

J08nY commented:
> IMO, ID of the certificate itself is fully determined by the contents from the module html page. What needs to get cleaned are various cert_ids detected as keywords in various sources: caveat text, pdf of the policy, etc. These point to other certificates, but are noisy and may point to Algorithms, or may just represent abitrary numbers, right?

Yes, this is what was done in the original code in a way. The list of "certlike matches" was cleaned by removing the algorithm numbers the certificate references, either from the webpage or from the policy in some table (that has "algo" in the header :))

> So the goal, prior to constructing the graph of references, would be to prune the _foreign_ cert_ids that are detected in `FIPSCertificate`. Does this make sense? Formally, there's no cert_id normalization or anything like that, you just delete detected foreign cert_ids that you consider weak in a sense.

Yes this is what was done and is I think important to keep it that way, otherwise the reference graph has way too many false positives.


> Also, I'll probably try to convince @GeorgeFI to take a look at building a graph of references, and also the transitive vulnerabilities. He did this for CC, so it makes sense to let him do this also here.
> 
> What do you think?

I don't think this is necessary, you can basically use `ReferenceFinder` the same way as it is (and as it is in CC), I actually sort of unified this part already, you just need to provide it with a source of cert_id references for a given graph. Originally there were two sources (and I think it makes sense to keep them), one from the page and one from the security policy. W.r.t. the transitive vulns computation, that should also be doable just with some renames to point to the correct clean cert_ids.

adamjanovsky commented:
@J08nY Ok, this is ready for review I suppose. Once we merge this, I intend to do #287, #231 and then release.

I checked that both CC and FIPS pipeline pass, tests pass, and notebooks are all executable and are producing "not-completely-off" plots. :D

Enjoy the review and don't be too harsh 🤞

J08nY commented on sec_certs/sample/fips.py:
```diff
@@ -1,206 +1,202 @@
         else:
-            html_items_found["vendor_www"] = vendor_string.get("href")
-            vendor_string = vendor_string.text.strip()
+            algorithms = set()
```
This line can move before the whole `if`.

J08nY commented on sec_certs/sample/fips.py:
```diff
@@ -1,206 +1,202 @@
+            [x.find("div", class_="col-md-3") for x in entries], [x.find("div", class_="col-md-9") for x in entries]
+        )
+        entries = [(FIPSHTMLParser.normalize_string(key.text), entry) for key, entry in entries]
+        entries = [parse_single_detail_entry(*x) for x in entries if x[0] in DETAILS_KEY_NORMALIZATION_DICT.keys()]
```
Hmm, we are losing some information here right? Or rather, that the check that `x[0] in DETAILS_KEY_NORMALIZATION_DICT.keys()` is required means that there is some field type on some cert page that is not in the `DETAILS_KEY_NORMALIZATION_DICT`. I think instead of the check we should do a run and log all of the additional details (along with their values) and then find a way to parse them as well.

adamjanovsky replied:
It's not difficult to parse additional info it we miss some (and we do). The question is to decide whether what we miss is worth parsing and storing. ATM we miss `Allowed Algorithms` which I don't find useful personally. 

I probably don't want to parse it all without some benefit analysis. I've prepared an issue: #289

J08nY commented on sec_certs/sample/fips.py:
```diff
@@ -420,354 +489,126 @@ def pandas_tuple(self) -> Tuple:
-                data = read_pdf(cert_file, pages="all" if all_pages else tables, silent=True)
+                tabular_data = read_pdf(cert.state.policy_pdf_path, pages=list(table_rich_page_numbers), silent=True)
+                cert.heuristics.algorithms |= set(
+                    itertools.chain.from_iterable([tables.get_algs_from_table(df.to_string()) for df in tabular_data])
```
This can be done without the list creation.

J08nY commented on sec_certs/sample/fips.py:
```diff
@@ -420,354 +489,126 @@ def pandas_tuple(self) -> Tuple:
-        matches.pop("#" + str(self.cert_id), None)
-        self.pdf_data.clean_cert_ids = matches
+        if self.pdf_data.keywords:
+            pdf_policy_ids = set(self.pdf_data.keywords["fips_cert_id"].get("Cert", dict().keys()))
```
The second argument to `get` being `dict().keys()` is very weird, what is it supposed to be?

adamjanovsky replied:
A bug. This should be `get("cert", dict()).keys()))` Thanks!

J08nY commented on sec_certs/sample/fips_algorithm.py:
```diff
@@ -1,49 +1,48 @@
-from typing import Optional
+from dataclasses import dataclass
+from datetime import date
+from typing import ClassVar, List, Tuple
```
Import annotations from future.

J08nY commented on sec_certs/utils/pandas.py:
```diff
@@ -9,7 +9,7 @@
 from pathlib import Path
 from shutil import copyfile
-from typing import Any, Dict, Final, List, Optional, Set, Tuple, Union
+from typing import Any, Final
```
This file does not have the annotations import from the future yet it was transformed to the Python >= 3.9 form of annotations. Was this manual or some pyupgrade issue? In any way, the import needs to be added otherwise this will likely break on Python 3.8.

J08nY commented on tests/data/cc/dataset/auxillary_datasets/maintenances/Maintenance updates.json:
```diff
@@ -0,0 +1,79 @@
+{
```
The filename of this dataset is quite different than of the other datasets in the tests (no snake case, space, ...).

J08nY commented on tests/data/settings_test.yaml:
```diff
@@ -1,59 +0,0 @@
----
```
The point of this test config was to disable tqdm display which messes with the output of some test runners in some consoles. I would like to keep this intact.

adamjanovsky replied:
Fixed with https://github.com/crocs-muni/sec-certs/pull/275/commits/4085c61d62e3e9d307ee8f05f3409d3cd10188ad

J08nY replied:
This is cool but this also needs to be setup to be loaded in the tests. Can we create some global pytest fixture that loads these settings as config? The previous test infra had this in some test's setUp methods. And I guess the test settings may need to be updated with the new defaults and property names (if that was not already done) and only the progress bar setting should differ from the default config.

adamjanovsky replied:
> This is cool but this also needs to be setup to be loaded in the tests. Can we create some global pytest fixture that loads these settings as config? The previous test infra had this in some test's setUp methods.

You're right, I'll fix that.

> And I guess the test settings may need to be updated with the new defaults and property names (if that was not already done) and only the progress bar setting should differ from the default config.

This should be reflected.

J08nY reviewed (changes requested):
What a great PR! I really like that this happened, it tied a lot of loose ends all over the codebase, especially in its darkest parts. I like the unification of the CLIs. The unification of dataset classes is also quite nice. Moving examples to notebooks is also a good call.

I also thing there are some things that should change in the PR:
 - The pyupgrade was done partially, or rather, some files are missing the `from __future__ import annotations` which enables the Python 3.9-type annotations on Python >= 3.7, which we need as we target 3.8. See the [pyupgrade readme](https://github.com/asottile/pyupgrade#pep-585-typing-rewrites) and [PEP 585](https://peps.python.org/pep-0585/).
 - I dislike that the `root_dir` property setter on dataset classes actually serializes the dataset. I think this is highly unintuitive and unexpected. I propose a more explicit solution in a comment in the review. The TL;DR; is to create a different method for moving/copying the dataset and make it the only method to change the root_dir.
 - It seems the whole PP dataset was included in the test data, was this intended? It shows up at almost 77k lines. I believe just downloading it from our site is a reasonable alternative. Although as it is already in git, removing it now is pointless (without a rebase and history editing the repository will have it forever).
 - We are not extracting some information from the FIPS module pages that we could (and I believe should). See the inline comment in [sec_certs/sample/fips.py](https://github.com/crocs-muni/sec-certs/pull/275/files/86a62cb7e212bfd8f956abcdeffaf9614f27891b#diff-4da23876e3eb5d55fdbbec5f078c600011284581b1ada95809141297317a9eb1)
 - Connected with the `root_dir` handling and serialization is the `DUMMY_PATH` issue. I see the current solution as an acceptable one, although I think there is room for improvement, we just need to identify the core issue in the API design and assumptions to see what needs to change. I think the current situation arises because our assumptions in different places in the code are not consistent or clear, specifically assumptions about whether a dataset object is backed by its JSON (directory structure) at different points in time as it is constructed from different sources. One possible way to resolve this is to say that the dataset object is always backed by its directory structure/JSON and explicitly require the path in all constructors, even the web ones. Other possible way is the opposite, to not assume the dataset is always backed, then disable the serialization features unless a path is set.

There are other suggestions in the code comments, but most are minor.

adamjanovsky commented on .pre-commit-config.yaml:
```diff
@@ -1,24 +1,29 @@
+    rev: v3.2.3
+    hooks:
+      - id: pyupgrade
+        args: ["--py38-plus"]
```
This is funny. There's bambillion of issues to introduce `--check` parameter to `pyupgrade`, all closed as "won't fix". See: https://github.com/asottile/pyupgrade/issues/578, https://github.com/asottile/pyupgrade/issues/460, https://github.com/asottile/pyupgrade/issues/442, https://github.com/asottile/pyupgrade/issues/310, https://github.com/asottile/pyupgrade/issues/210. 

**But**,  while the contents are changed, if they are, `pyupgrade` exits with nonzero status. Ergo the pre-commit won't pass and you can revisit the changes yourself and revert them if necessary. 

Also, pyupgrade cannot clean imports that are no longer unncesseary. E.g., if you rewrite `typing.List` to `list`, then `import typing` will be left there and Flake8 will complain. Also just noting :).

adamjanovsky commented:
> The pyupgrade was done partially, or rather, some files are missing the from __future__ import annotations which enables the Python 3.9-type annotations on Python >= 3.7, which we need as we target 3.8. See the [pyupgrade readme](https://github.com/asottile/pyupgrade#pep-585-typing-rewrites) and [PEP 585](https://peps.python.org/pep-0585/).

I see, this is because `pyupgrade` only works on files that have `from __future__ import annotations` at the top. I've added a [flake8 plugin](https://pypi.org/project/flake8-future-annotations/) that checks for its absence so that this doesn't repeat in the future. 

I've added these import statements and upgraded the rest of the codebase with https://github.com/crocs-muni/sec-certs/pull/275/commits/6ce7007ab5b5f721abf6d8659ef4bd54e4f61590

adamjanovsky commented on Dockerfile:
```diff
@@ -51,9 +50,6 @@ RUN \
   pip3 install -e .
 
-# Download spacy language model
-RUN python3 -m spacy download en_core_web_sm
```
Yes and yes. The model is now implicitly downloaded via [requirements.in](https://github.com/crocs-muni/sec-certs/blob/6ce7007ab5b5f721abf6d8659ef4bd54e4f61590/requirements/requirements.in#L25). It saves one line during installation of the tool that was very easy to forgot. Previously

```
pip3 install sec-certs
python3 -m spacy download en_core_web_sm
```

now just `pip3 install sec-certs`

adamjanovsky commented on docs/installation.md:
````diff
@@ -7,7 +7,6 @@ The tool can be installed from PyPi with
 
 ```bash
 pip install -U sec-certs
-python3 -m spacy download en_core_web_sm
````
See https://github.com/crocs-muni/sec-certs/pull/275#discussion_r1039693198

adamjanovsky commented on requirements/requirements.in:
```diff
@@ -23,8 +22,9 @@ setuptools-scm
 ipykernel
 ipywidgets
 spacy
+en_core_web_sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.4.1/en_core_web_sm-3.4.1.tar.gz
```
I also reacted to your previous notes as I'm processing your review the same way you've processed the PR, immediate dispatch 😄 . 

I see what you're saying. I think this will bring less trouble than the previous solution. If it breaks, you should now immediately. At that moment you can always fetch the newest model without reinstalling sec-certs or anything like that.

adamjanovsky commented on sec_certs/__main__.py:
```diff
@@ -0,0 +1,3 @@
+from sec_certs.cli import main
+
+main()
```
> Huh, this is interesting, I haven't seen this before. I guess it works but I would still guard it with if __name__ == "__main__", unless that is not possible and this needs to also exec on import or something.

Thanks, I'll add the guard

> Actually, looking at the entrypoints setup in setup.py this is unused right?

Well, that depends. This enables to run `sec-certs` as a module with `python -m sec-certs`.

J08nY replied:
Ah I see! This is cool, keep it then!

adamjanovsky replied:
Entry guard and some super symple exit codes 1/0 added with https://github.com/crocs-muni/sec-certs/pull/275/commits/8f7a14bb905b7d00570b2535f6faa5a19cd0875b

J08nY commented on Dockerfile:
```diff
@@ -18,7 +18,6 @@ RUN DEBIAN_FRONTEND="noninteractive" apt-get -y install tzdata
 RUN apt-get install build-essential libpoppler-cpp-dev pkg-config python3-dev -y
 RUN apt-get install libqpdf-dev -y
 RUN apt-get install default-jdk -y
-RUN apt-get install graphviz -y
```
Ok, makes sense.

adamjanovsky commented on sec_certs/constants.py:
```diff
@@ -1,4 +1,7 @@
 import re
+from pathlib import Path
+
+DUMMY_NONEXISTING_PATH = Path("/this/is/dummy/nonexisting/path")
```
😆 we could generate that variable at runtime with data from `/dev/random` :D

adamjanovsky commented on sec_certs/dataset/fips.py:
```diff
@@ -1,572 +1,339 @@
+    def algorithm_dataset_path(self) -> Path:
+        return self.algorithms_dir / "algorithms.json"
+
+    def __getitem__(self, item: str) -> FIPSCertificate:
```
Previously they were not. I didn' care, but I wanted to access them by `dset["1"]` which was not possible. This is a catdog, you're right. I don't see a reason that prevents us from indexing them by `cert_id`. What do you think?

J08nY replied:
Ah I see. What about getting to them by:
```
dset[5]
dset["12deadbeef"]
```
and doing a type test on int vs str?

I think that is the cleanest solution. Btw. the page probably depends on indexing certs in the dataset by the digest so I would like to keep it that way.

adamjanovsky replied:
Well, on all other places the `cert_id` (the short identifier) is a string, not int. :( Not sure what to do...

J08nY replied:
What about:
```
try:
    return super().__getitem__(item)
except:
    return super().__getitem__(fips_dgst(item))
```
Or something of that sort? Gets rid of the 5 magic number, though the exceptions might get worse quality.

adamjanovsky replied:
👍

J08nY commented on sec_certs/dataset/common_criteria.py:
```diff
@@ -752,13 +643,13 @@ def _extract_targets_frontpage(self, fresh: bool = True) -> None:
-    def _extract_report_keywords(self, fresh: bool = True) -> None:
-        certs_to_process = [x for x in self if x.state.report_is_ok_to_analyze(fresh)]
+    def _extract_report_keywords(self) -> None:
+        logger.info("Extracting report keywords")
```
Done in #288.

adamjanovsky commented on sec_certs/dataset/cpe.py:
```diff
@@ -15,33 +14,44 @@
-    init_lookup_dicts: InitVar[bool] = True
-    cpe_xml_basename: ClassVar[str] = "official-cpe-dictionary_v2.3.xml"
-    cpe_url: ClassVar[str] = "https://nvd.nist.gov/feeds/xml/cpe/dictionary/" + cpe_xml_basename + ".zip"
+    CPE_XML_BASENAME: ClassVar[str] = "official-cpe-dictionary_v2.3.xml"
```
won't fix based on our personal discussion.

adamjanovsky commented on sec_certs/dataset/cve.py:
```diff
@@ -18,18 +17,30 @@
-    cpe_to_cve_ids_lookup: Dict[str, Set[str]] = field(init=False)
-    cve_url: Final[str] = "https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-"
-    cpe_match_feed_url: Final[str] = "https://nvd.nist.gov/feeds/json/cpematch/1.0/nvdcpematch-1.0.json.zip"
+    CVE_URL: ClassVar[str] = "https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-"
```
won't fix based on our personal discussion.

adamjanovsky commented:
> Connected with the root_dir handling and serialization is the DUMMY_PATH issue. I see the current solution as an acceptable one, although I think there is room for improvement, we just need to identify the core issue in the API design and assumptions to see what needs to change. I think the current situation arises because our assumptions in different places in the code are not consistent or clear, specifically assumptions about whether a dataset object is backed by its JSON (directory structure) at different points in time as it is constructed from different sources.

You're perfectly right about the cause. 

>  One possible way to resolve this is to say that the dataset object is always backed by its directory structure/JSON and explicitly require the path in all constructors, even the web ones. Other possible way is the opposite, to not assume the dataset is always backed, then disable the serialization features unless a path is set.

I was actually opting for the second, which is kind of the current state. The serialization will fail unless a path is set. See

https://github.com/crocs-muni/sec-certs/blob/f14dfe3900baa0a87a51f42188646f409b9beeb9/sec_certs/serialization/json.py#L50-L51

It's just implemented with that `DUMMY_NONEXISTING_PATH` variable...

adamjanovsky commented:
> I dislike that the root_dir property setter on dataset classes actually serializes the dataset. I think this is highly unintuitive and unexpected. I propose a more explicit solution in a comment in the review. The TL;DR; is to create a different method for moving/copying the dataset and make it the only method to change the root_dir.

I agree this can be counterintuitive. I'll take a look at that, thanks for spotting this.

adamjanovsky commented on sec_certs/dataset/dataset.py:
```diff
@@ -1,73 +1,96 @@
+        if old_dset.root_dir != self.root_dir:
+            logger.info(f"Changing root dir of partially processed dataset. All contents will get copied to {new_dir}")
+            self._copy_dataset_contents(old_dset)
+            self.to_json()
```
Agree, I'll do it this way.

adamjanovsky commented on sec_certs/dataset/dataset.py:
```diff
@@ -120,8 +150,16 @@ def __eq__(self, other: object) -> bool:
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            dset_path = Path(tmp_dir) / filename
+            helpers.download_file(url, dset_path, show_progress_bar=True, progress_bar_desc=progress_bar_desc)
+            return cls.from_json(dset_path)
```
And it dies right after the return, right? I'll just set the root dir to `DUMMY_NONEXISTING_PATH` then 😆

adamjanovsky commented on sec_certs/dataset/fips.py:
```diff
@@ -1,572 +1,339 @@
+    def _extract_data(self) -> None:
+        logger.info("Extracting various data from certification artifacts.")
+        for cert in self:
+            cert.state.policy_extract_ok = True
```
Well the status is set to ok and then updated to False if it fails. Certs cannot change it, because they don't know whether the processing was invoked only on the failed certificates (in that cases, certs with `something_extract_ok = True` are discarded).

adamjanovsky commented on sec_certs/sample/cpe.py:
```diff
@@ -26,8 +26,6 @@ class CPE(PandasSerializableType, ComplexSerializableType):
         "item_name",
         "version",
         "title",
-        "start_version",
-        "end_version",
```
I can't imagine why would anyone want to use this from pandas. It's not even a CPE thing. We still support it during CPE matching, but we don't use that anymore.

J08nY replied:
Ah, it was removed from pandas columns I thought it was removed from serialized attrs.

J08nY reviewed (commented):
lgtm except the hash impl.

adamjanovsky commented on sec_certs/dataset/dataset.py:
```diff
@@ -1,73 +1,92 @@
 
     @root_dir.setter
-    def root_dir(self, new_dir: Union[str, Path]) -> None:
+    def root_dir(self: DatasetSubType, new_dir: str | Path) -> None:
```
> Should this be a public setter and not a function like _set_root_dir? It allows for getting the dataset into a state where its root_dir points to wherever, which might not even exist. Thus I think this should be more private and only used internally. But maybe we want users to be able to set this.

well if we allow the users to instantiate dataset without a `root_dir`, they need to set it afterwards somehow. So it must be public. The design right now is that you can use the dataset as a datastructure without any explicit paths. As soon as you want to process it somehow, it will scream that you must set `root_dir`. 

> Another thing is that this tests whether the root_dir is a file but not whether it is actually a directory. I understand that we may not want to check that, as it may not (yet) be a directory when this gets called but still, checking that it is not a file seems extra and a weird check. But this is probably just a matter of expectations. I would not expect such a check here.

I don't mind if it doesn't exist. I just mind if it's a file. Then creating a directory on the same path would overwrite it.

J08nY replied:
I see, makes sense.

adamjanovsky commented on sec_certs/constants.py:
```diff
@@ -1,4 +1,7 @@
 import re
+from pathlib import Path
+
+DUMMY_NONEXISTING_PATH = Path("/this/is/dummy/nonexisting/path")
```
Either this or bambillion of MyPy checks, because the alternative is setting `root_dir = None`, which would break a lot of things.

adamjanovsky replied:
Created #290 to address this.

adamjanovsky commented on sec_certs/sample/certificate.py:
```diff
@@ -56,19 +76,22 @@ def __eq__(self, other: object) -> bool:
 
-    def to_dict(self) -> Dict[str, Any]:
+    def __hash__(self) -> int:
+        return hash(self.dgst)
```
Hm, this is problematic then. We want two things:

- Have `Certificate` objects comparable using `==` operator (so that, e.g., dictionaries carrying `Certificate`s can be compared as well, which assures that `Dataset` objects can be compared)
- Have `Certificate` objects *hashable* so that we can use them in sets. 

Since the classes define mutable attributes, there's a direct conflict between these two requirements?

J08nY replied:
Yes, there is a conflict. But the second thing is in conflict with the class being mutable by itself, the conflict is not between the two requirements.

adamjanovsky replied:
Ok, so only `FIPSCertificate` needed to be hashable and it no longer does. So we're good.
